certloader/certstore_enabled.go

```diff
@@ -66,33 +67,57 @@ func (c *certstoreCertificate) Reload() error {
 		return err
 	}
 
-	var certAndKey *tls.Certificate
+	// filter any certificates with the matching name, as the keychain allows
+	// multiple certificates with the same name
+	var candidates []certstore.Identity
 	for _, identity := range identitites {
 		chain, err := identity.CertificateChain()
 		if err != nil {
 			continue
 		}
 
-		signer, err := identity.Signer()
+		if chain[0].Subject.CommonName == c.commonName {
+			candidates = append(candidates, identity)
+		}
+	}
+
+	if len(candidates) == 0 {
+		return fmt.Errorf("unable to find identity with common name '%s' in keychain", c.commonName)
+	}
+
+	// sort the candidates by ascending NotAfter
+	sort.Slice(candidates, func(i, j int) bool {
+		i_chain, err := candidates[i].CertificateChain()
 		if err != nil {
-			continue
+			return true
 		}
 
-		if chain[0].Subject.CommonName == c.commonName {
-			certAndKey = &tls.Certificate{
-				Certificate: serializeChain(chain),
-				PrivateKey:  signer,
-			}
-			break
+		j_chain, err := candidates[j].CertificateChain()
+		if err != nil {
+			return false
 		}
+
+		return i_chain[0].NotAfter.Before(j_chain[0].NotAfter)
+	})
+
+	// choose the certificate with the NotAfter furthest in the future
+	chosenIdentity := candidates[len(candidates)-1]
+	chain, err := chosenIdentity.CertificateChain()
+	if err != nil {
+		return fmt.Errorf("unable to find identity with common name '%s' in keychain", c.commonName)
+	}
+	signer, err := chosenIdentity.Signer()
+	if err != nil {
+		return fmt.Errorf("unable to find identity with common name '%s' in keychain", c.commonName)
 	}
 
-	if certAndKey != nil {
-		atomic.StorePointer(&c.cached, unsafe.Pointer(certAndKey))
-		return nil
+	certAndKey := &tls.Certificate{
+		Certificate: serializeChain(chain),
+		PrivateKey:  signer,
```

Comment:
While we're at it, let's set `Leaf: chain[0]` here (assuming elem 0 is the leaf).